src/osd/ReplicatedPG.cc

```diff
@@ -1496,6 +1496,22 @@ void ReplicatedPG::do_op(OpRequestRef& op)
     return;
   }
 
+  // discard due to cluster full transition?  (we discard any op that
+  // originates before the cluster or pool is marked full; the client
+  // will resend after the full flag is removed or if they expect the
+  // op to succeed despite being full).
+  // FIXME: we exclude mds writes for now.
+  if (!m->get_source().is_mds() &&
+      info.history.last_epoch_marked_full > m->get_map_epoch()) {
+    dout(10) << __func__ << " discarding op sent before full " << m << " "
+	     << *m << dendl;
+    return;
+  }
+  if (osd->check_failsafe_full()) {
+    osd->reply_op_error(op, -ENOSPC);
```

Comment:
My thought is that in the cluster full or pool full case, the client knows about it, and expects us to drop any requests sent before that epoch.

If the client opts to send something *while* the cluster is full, it is hopefully doing so because it expects the op to succeed anyway (e.g., delete).. but if it *doesn't* succeed, it should get an explicit error.

Similarly, if we hit the failsafe, we are failing the write in a way that the client can't predict, and it needs to know about it.  Maybe the OSD will go down soon anyway, but I don't think that's guaranteed to be the case...?  (If it is, then we could drop the op in that case too)